exchanges/huobi/huobi_futures.go

```diff
@@ -1249,6 +1249,11 @@ func (h *HUOBI) formatFuturesPair(p currency.Pair, convertQuoteToExpiry bool) (s
 // We need this because some apis, such as ticker, use BTC_CW, NW, CQ, NQ
 // Other apis, such as contract_info, use contract type of this_week, next_week, quarter (sic), and next_quater
 func (h *HUOBI) pairFromContractExpiryCode(p currency.Pair) (currency.Pair, error) {
+	if h.futureContractCodes == nil { // Worst race-case here is that we fetch twice. No avoiding it.
```

Comment:
A nil check is not race proof
```
WARNING: DATA RACE
Read at 0x0001014ee5a0 by goroutine 238:
  github.com/thrasher-corp/gocryptotrader/exchanges/huobi.(*HUOBI).pairFromContractExpiryCode()
      /Users/GRASS/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/huobi/huobi_futures.go:1252 +0x64
  github.com/thrasher-corp/gocryptotrader/exchanges/huobi.TestButts.func1()
      /Users/GRASS/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/huobi/huobi_test.go:2830 +0x11c
  github.com/thrasher-corp/gocryptotrader/exchanges/huobi.TestButts.gowrap1()
      /Users/GRASS/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/huobi/huobi_test.go:2834 +0x44

Previous write at 0x0001014ee5a0 by goroutine 239:
  github.com/thrasher-corp/gocryptotrader/exchanges/huobi.(*HUOBI).FetchTradablePairs()
      /Users/GRASS/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/huobi/huobi_wrapper.go:292 +0x1ec
  github.com/thrasher-corp/gocryptotrader/exchanges/huobi.TestButts.func2()
      /Users/GRASS/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/huobi/huobi_test.go:2837 +0x8c
  github.com/thrasher-corp/gocryptotrader/exchanges/huobi.TestButts.gowrap2()
      /Users/GRASS/go/src/github.com/thrasher-corp/gocryptotrader/exchanges/huobi/huobi_test.go:2838 +0x44
```
You can argue that me calling both this and `FetchTradablePairs` in go routines to prove this point isn't the best behaviour, but I'd say the same about not synchronising tradable pairs for futures with expiring contracts 😄

Reply:
Nit pick: Race proof versus Race detected are not the same, though, right?
I think that a nil check is race proof and atomic.
The access isn't synchronised, but it is atomic.

Suppose we wrap a read lock around this nil check.
We then have to unlock the read lock to call fetchPairs. That'd be more of a race opportunity there, than just straight-lining from the nil check into the fetch call.

All of that said, if this has the potential to trigger the race detector I'll work a way around it somehow.

Reply:
Fixed 1ea8fd4df ( with extreme prejudice 😈 )